app/models/service.rb

```diff
@@ -46,4 +46,40 @@ def required_tiers(tier_name)
   def to_param
     slug
   end
+
+  def duplicate!(new_lgsl_code)
+    new_service = dup
+    new_service.lgsl_code = new_lgsl_code
+    new_service.label = "TRANSITION -- #{label}"
+    new_service.slug = "TRANSITION -- #{slug}"
+    new_service.save!
+
+    service_tiers.each do |service_tier|
+      new_service_tier = service_tier.dup
+      new_service_tier.service_id = new_service.id
+      new_service_tier.save!
+    end
+
+    service_interactions.each do |service_interaction|
+      new_service_interaction = service_interaction.dup
+      new_service_interaction.service_id = new_service.id
+      new_service_interaction.save!
+
+      service_interaction.links.each do |link|
+        new_link = link.dup
+        new_link.service_interaction_id = new_service_interaction.id
+        new_link.save!
+      end
+    end
```

Comment:
I'm being nit-picky for temporary code but you can make this code a bit shorter by relying on relationship autosaving:

```
  new_service = dup
  new_service.assign_attributes(
    lgsl_code: lgsl_code,
    label: "TRANSITION -- #{label}",
    slug: "TRANSITION -- #{slug}",
  )
  new_service.save!
  
  new_service.service_interactions = service_interactions.map do |si|
    si.dup.tap { |new_si| new_si.links = si.links.map(&:dup) }
  end
  new_service.service_tiers = service_tiers.map(&:dup)
```

Reply:
Oh and I also think we should also put it in a transaction so it's all or nothing if that are any problems.

Reply:
Thanks - fixed up as per your suggestion